### PR TITLE
Detect outliers using LOF

### PR DESCRIPTION
**Pull request by urschrei**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/main/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

Still draft due to:
- <s>perf: two of the LRD passes look very similar. I need to figure out whether this can be tidied up, or whether it's ultimately not going to affect perf much</s> Can't think of a particularly more elegant way to do this
- Tests, tests, tests
- Other geometries / collections?

### CODE REVIEW

File: geo/src/algorithm/outlier_detection.rs
Comment (urschrei) [inline geo/src/algorithm/outlier_detection.rs:135]: I suspect there's a way to calculate LRDs for a subset ( the point's neighbours) as we go, but I can't get it right so we're stuck with `local_reachability_densities` and a redundant `kth_dist` calculation for now. I think it would ultimately end up being two passes over the data anyway, but it might use less space.

**Comment by urschrei** [comment]
The paper notes that in order to calculate outliers in a robust fashion it might be necessary to calculate an "ensemble" using a range of nearest-neighour values. Since the r*-tree is expensive to fill and can be re-used, it would make sense to have a "prepared outlier calculator" API that holds on to the tree, allowing you to pass in a new `k_neighbours` value and getting a re-calculated dataset back out – a dataset is guaranteed to have the same dimensions (one `LofScore` entry per input point), so it will be easy to either aggregate scores once you've run it over a range of input neighbour sizes, or calculate them as you go in order to save space if you know what aggregate measure will give you good results and you want to save space.

All suggestions for an API gratefully received.

**Comment by urschrei** [comment]
> The paper notes that in order to calculate outliers in a robust fashion it might be necessary to calculate an "ensemble" using a range of nearest-neighour values. Since the r*-tree is expensive to fill and can be re-used, it would make sense to have a "prepared outlier calculator" API that holds on to the tree, allowing you to pass in a new `k_neighbours` value and getting a re-calculated dataset back out – a dataset is guaranteed to have the same dimensions (one `LofScore` entry per input point), so it will be easy to either aggregate scores once you've run it over a range of input neighbour sizes

This is done!

File: geo/src/algorithm/outlier_detection.rs
Comment (michaelkirk) [inline geo/src/algorithm/outlier_detection.rs:203]:
```suggestion
{
    assert!(kneighbours > 0);
```

Would give a better panic message than the later `unwrap` panic.

Also, I'm not sure the meaning of this algo when kneighbors == 1, but maybe that's fine and not worth panicking over.

File: geo/src/algorithm/outlier_detection.rs
Comment (michaelkirk) [inline geo/src/algorithm/outlier_detection.rs:234]: It seems like we'd get a div-by-0 here when there are sufficient repeated points. Does the algorithm provide sensible results in that case?

File: geo/src/algorithm/outlier_detection.rs
Comment (michaelkirk) [inline geo/src/algorithm/outlier_detection.rs:66]: Nice! Maybe mention in the docs that you'd want to do this for efficiency sake. It's implied by use of the word "prepared" if you know what that means, but is that kind of obscure?

Maybe something that mentions "efficiency" like:
>  /// A PreparedDetector can efficiently recompute outliers with different k_neigbhors value.

File: geo/src/algorithm/outlier_detection.rs
Comment (urschrei) [reply geo/src/algorithm/outlier_detection.rs:234]: It returns either infinity or `NaN` at the moment (which is documented). I'm not sure how we can avoid that without affecting performance, though…

File: geo/src/algorithm/outlier_detection.rs
Comment (urschrei) [reply geo/src/algorithm/outlier_detection.rs:203]: 1 will just give lower-quality results. Note that passing `kneighbours = 0` won't cause a panic down the line - we immediately return all the LOF scores set to 1, which is essentially saying "we couldn't calculate them". I've added a `debug_assert` – dunno if that's good practice though.

Comment (michaelkirk) [reply geo/src/algorithm/outlier_detection.rs:208]: Hmm, when I run this:
```
        let v = [
            Point::new(0.0, 1.0),
            Point::new(0.0, 2.0),
            Point::new(3.0, 1.0),
            Point::new(1.0, 1.0),
        ];

        let scores = v.outliers(0);
```

I see this panic:
```
called `Option::unwrap()` on a `None` value
thread 'algorithm::outlier_detection::tests::test_same_points' panicked at 'called `Option::unwrap()` on a `None` value', geo/src/algorithm/outlier_detection.rs:232:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:48:5
   3: core::option::Option<T>::unwrap
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/option.rs:775:21
   4: geo::algorithm::outlier_detection::lof::{{closure}}
             at ./src/algorithm/outlier_detection.rs:228:28
   5: core::iter::adapters::map::map_fold::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/map.rs:84:28
   6: core::iter::traits::iterator::Iterator::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:2370:21
   7: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/map.rs:124:9
   8: core::iter::traits::iterator::Iterator::for_each
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:787:9
   9: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/alloc/src/vec/spec_extend.rs:40:17
  10: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/alloc/src/vec/spec_from_iter_nested.rs:62:9
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/alloc/src/vec/spec_from_iter.rs:33:9
  12: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/alloc/src/vec/mod.rs:2645:9
  13: core::iter::traits::iterator::Iterator::collect
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:1792:9
  14: geo::algorithm::outlier_detection::lof
             at ./src/algorithm/outlier_detection.rs:224:48
  15: geo::algorithm::outlier_detection::PreparedDetector<T>::outliers
             at ./src/algorithm/outlier_detection.rs:192:9
  16: <[geo_types::geometry::point::Point<T>] as geo::algorithm::outlier_detection::OutlierDetection<T>>::outliers
             at ./src/algorithm/outlier_detection.rs:299:9
  17: geo::algorithm::outlier_detection::tests::test_same_points
             at ./src/algorithm/outlier_detection.rs:429:22
  18: geo::algorithm::outlier_detection::tests::test_same_points::{{closure}}
             at ./src/algorithm/outlier_detection.rs:420:5
  19: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
  20: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

File: geo/src/algorithm/outlier_detection.rs
Comment (urschrei) [reply geo/src/algorithm/outlier_detection.rs:80]: Just to clarify:
Are you suggesting that instead of an outlier score (and attendant custom type), by default we just return a `Vec<T>` and if someone wants the indices they can do `self.outliers(neighbor_count).iter().enumerated({ |(score, idx)|  ...  })`?

File: geo/src/algorithm/outlier_detection.rs
Comment (urschrei) [reply geo/src/algorithm/outlier_detection.rs:208]: Ugh yes my fault: we only check that neighbours < points. But we could just extend that check to neighbours > 0 instead of just panicking? Which is better: indeterminate result, or a panic?

File: geo/src/algorithm/outlier_detection.rs
Comment (michaelkirk) [reply geo/src/algorithm/outlier_detection.rs:80]: Sorry, I see now that my example code had a bug - it should have been `Vec<T>`, not just a single `T`.

So, yes! You've understood what I was suggesting. Do you think it's an improvement?

File: geo/src/algorithm/outlier_detection.rs
Comment (michaelkirk) [reply geo/src/algorithm/outlier_detection.rs:208]: Neighbors less than 0 seems almost certainly to be a programmer error, so I'm very biased for panicing in debug builds.

Doing our best to avoid panics in release builds seems nice though, so I have a slight bias in favor of "blundering on" there.

Comment (urschrei) [reply geo/src/algorithm/outlier_detection.rs:234]: We now panic in release, and return 1 in release if `neighbours > points` OR `neighbours < 1`.

File: geo/src/algorithm/outlier_detection.rs
Comment (urschrei) [reply geo/src/algorithm/outlier_detection.rs:263]: I've made this possible result extra-clear in the docs, and because this input can't produce a panic I'm OK with it as is.

File: geo/src/algorithm/outlier_detection.rs
Comment (michaelkirk) [reply geo/src/algorithm/outlier_detection.rs:235]: > We now panic in release,

We now panic in ~~release~~ debug,

I think this was just a typo in your github comment. The code looks right -thank you for making this change!

File: geo/src/algorithm/outlier_detection.rs
Comment (michaelkirk) [inline geo/src/algorithm/outlier_detection.rs:110]: You changed this to `outlier_scores` in response to my feedback, and then changed it back to `outliers`. I just wanted to make sure that was intentional.

Comment (urschrei) [reply geo/src/algorithm/outlier_detection.rs:109]: Yep! Sorry.

**Comment by urschrei** [comment]
All review comments addressed, I think.

**Review by michaelkirk — APPROVED** [review]
LGTM! Not sure if you were looking for more reviews, so I'll leave it to you if you want to merge it.

**Comment by urschrei** [comment]
Thanks for the review! I think it's in good shape now but I'll leave it open for a couple of days / until @frewsxcv has kicked the tires on it.

**Pull request merged**
